# 5 - User-Defined Functions and Good Programming Style<br>*(Ch. 2.6 and Ch. 2.7)*

## Many times we would like to have some code that we can use over and over again without copy/pasting it every time.

To do this we can write a *function*. (Like the built-in functions that are loaded by Python when it starts.)

The basic syntax is:

In [ ]:
# Example 1
def sample_function1():
    print("A function just executed, oh my!")

sample_function1()

# Example 2
def sample_function2(a,b):
    
    return a,b  #a**b

y1,y2 = sample_function2(3,4)
print(y1)
print(y2)

## Example: 
Define a function that finds the hypoteneuse of a right triangle given the other two lengths. Test it for side lengths $x=3$ and $y=4$.

In [ ]:
def hyp(s1,s2):
    return (s1**2+s2**2)**0.5

s1 = 8
s2 = 6
print('Hypotenuse: {0}'.format(hyp(s1,s2)))

<br><br><br><br><br><br><br><br><br><br><br><br>
### Solution:

In [ ]:
def hypot(x,y):
    """Determines the hypotenuse of a right triangle with the given sides.
    Use in the following format: h = hypot(x,y)"""
    return (x**2+y**2)**0.5

h = hypot(3,4)
print(h)

print(float.__doc__)

What does that last line do? We'll get to that in just a moment.
<br><br><br><br><br><br><br><br><br><br><br><br><br><br>

## Basic rules for using functions:

* A function must be defined before it is used.

    + Option 1: It is common practice to define all of your functions at the top of a .py file.
    
    + Option 2: If your user-defined function (e.g. `myfunction`) will be used by many unrelated programs, you can write all such functions inside a single file (e.g. `thuecksfns.py`) and then use `from thuecksfns.py import myfunction` as if you were importing a function from a module.
    
* When executing (***calling***) a function, all of the code inside the function is executed once.

* The special keyword `return` immediately terminates the function (similar to `break` in a loop).

    + This can be used to assign the output of some function to an object, as in the above example.
    
    + If the function never executes a `return` statement, it returns `None` at the end of the code; we generally just ignore this (as in sample_function1() above).
    
* All variables created inside the definition of a function are ***local variables*** and exist only within that function. Once the function completes, local variables disappear.



### Exercise:

Check the following code, determine why it doesn't work, and devise a fix.


In [1]:
balance = 100
def add_interest(balance, rate):
    balance += balance * rate / 100

for year in range(4):
    add_interest(balance,5)
    print("Balance after year {0}: ${1:.2f}".format(year+1,balance))

Balance after year 1: $100.00
Balance after year 2: $100.00
Balance after year 3: $100.00
Balance after year 4: $100.00


<br><br><br><br><br><br><br><br><br><br><br><br>

#### Pre-written solution:

In [2]:
balance = 100
def add_interest(bal, rate):
    return bal * rate / 100

for year in range(4):
    balance += add_interest(balance,5)
    print("Balance after year {0}: ${1:.2f}".format(year+1,balance))

Balance after year 1: $105.00
Balance after year 2: $110.25
Balance after year 3: $115.76
Balance after year 4: $121.55


<br><br><br><br><br><br><br><br><br><br><br><br><br><br>

## Docstrings are special strings that summarize the purpose of a function.

* The docstring is defined on the first line below the definition (wrapped in triple quotes).

* It is then accessible with the `__doc__` method.

* For anything beyond the simplest functions, always include a docstring! (You'll thank yourself when you come back to your code weeks after writing it.)

### Previous example:

In [ ]:
def hypot(x,y):
    """Determines the hypotenuse of a right triangle with the given sides."""
    return (x**2+y**2)**0.5

h = hypot(3,4)
print(h)

print(hypot.__doc__)

<br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br>
## Exercise 2.11: Binomial coefficients
The binomial coefficient ${n\choose k}$ is an integer equal to
$$ {n\choose k} = {n!\over k!(n-k)!}
  = {n\times(n-1)\times(n-2)\times\ldots\times(n-k+1)\over
     1\times2\times\ldots\times k}
$$
when $k\ge1$, or ${n\choose0}=1$ when $k=0$.

a. Using this form for the binomial coefficient, write a user-defined
  function `binomial(n,k)` that calculates the binomial coefficient
  for given $n$ and $k$.  Make sure your function returns the answer in the
  form of an integer (not a float) and gives the correct value of 1 for the
  case where $k=0$. *Hint: The last expression above has a numerator and denominator that each have k terms.*
  
b. Using your function write a program to print out the first 20 lines
  of ``Pascal's triangle.''  The $n$th line of Pascal's triangle contains
  $n+1$ numbers, which are the coefficients ${n\choose 0}$, ${n\choose1}$,
  and so on up to ${n\choose n}$.  Thus the first few lines are

1 1

1 2 1

1 3 3 1

1 4 6 4 1

***Hint: You can replace the automatic 'end-of-line' at the end of a print statement and replace it with a space using a command like so:***
```python
print('This is an',end=' ')
print('incomplete sentence that is now complete.')
```

c. The probability that an unbiased coin, tossed $n$ times, will come up
  heads $k$ times is ${n\choose k}/2^n$.  Write a program to calculate
  (a) the total probability that a coin tossed 100 times comes up heads
  exactly 60 times, and (b) the probability that it comes up heads 60 or
  more times.


In [10]:
def binomial(n,k):
    num = 1
    den = 1
    for m in range(k):
        num *= n-m
        den *= m+1
    return num//den

for n in range(20):
    for k in range(n+1):
        print(binomial(n,k), end=' ')
    print() # Start new line
    

1 
1 1 
1 2 1 
1 3 3 1 
1 4 6 4 1 
1 5 10 10 5 1 
1 6 15 20 15 6 1 
1 7 21 35 35 21 7 1 
1 8 28 56 70 56 28 8 1 
1 9 36 84 126 126 84 36 9 1 
1 10 45 120 210 252 210 120 45 10 1 
1 11 55 165 330 462 462 330 165 55 11 1 
1 12 66 220 495 792 924 792 495 220 66 12 1 
1 13 78 286 715 1287 1716 1716 1287 715 286 78 13 1 
1 14 91 364 1001 2002 3003 3432 3003 2002 1001 364 91 14 1 
1 15 105 455 1365 3003 5005 6435 6435 5005 3003 1365 455 105 15 1 
1 16 120 560 1820 4368 8008 11440 12870 11440 8008 4368 1820 560 120 16 1 
1 17 136 680 2380 6188 12376 19448 24310 24310 19448 12376 6188 2380 680 136 17 1 
1 18 153 816 3060 8568 18564 31824 43758 48620 43758 31824 18564 8568 3060 816 153 18 1 
1 19 171 969 3876 11628 27132 50388 75582 92378 92378 75582 50388 27132 11628 3876 969 171 19 1 


<br><br><br><br><br><br><br><br><br><br>
Solution:

In [ ]:
def binomial(n,k):
    p1 = 1
    p2 = 1
    for m in range(k):
        p1 *= n - m
        p2 *= m + 1
    return p1//p2

for n in range(20):
    for k in range(n+1):
        print(binomial(n,k),end=' ') #The " end=' ' " command prevents a new line at the end of the print statement.
    print() # Start a new line

In [12]:
import numpy as np
np.sin(2*np.pi)

-2.4492935982947064e-16

<br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br>

## The *arguments* of a function can be treated in multiple ways.

**Functions with arguments can be treated in a *positional* manner:**

```python
def f(a,b,c):
   print(a,b,c)

# Arguments passed to the function are assigned in order
# to the arguments in the function definition.
f(1,2,3)
f(2,1,3)
```

**Functions with arguments can be treated in a *keyword* manner:**

```python
def f(a,b,c):
   print(a,b,c)

# Being explicit means order does not matter.
f(a=1,b=2,c=3)
f(c=3,a=1,b=2)
```

**Functions can be defined with *default arguments*:**

```python
def f(a=1,b=2,c=3):
    print(a,b,c)

f(b=2,c=3)      # `a` reverts to its default value
f(2,3,4)        # positional
f(a=3,c=4,b=3)  # keyword
f(5,b=2,c=4)    # an OK mix (positional then keyword)
f(b=2,5,6)      # ERROR - can't do keyword then positional
```

In [19]:
def f(a,b,c='end', d='no', e='yes', f='bye'):
   print(a,b,c,d,e,f)

f(1,2, e='ok')




1 2 end no ok bye


<br><br><br><br><br><br><br><br><br><br><br><br><br><br>
## Example: Exercise 2.13 - Recursion
A useful feature of user-defined functions is *recursion*, the
ability of a function to call itself.  For example, consider the following
definition of the factorial $n!$ of a positive integer $n$:
$$n! = \biggl\lbrace\begin{array}{ll}
  1 & \qquad\mbox{if $n=1$,} \\
  n\times(n-1)! & \qquad\mbox{if $n>1$.}
\end{array}$$
This constitutes a complete definition of the factorial which allows us to
calculate the value of $n!$ for any positive integer.  We can employ this
definition directly to create a Python function for factorials, like this:
```python
def factorial(n):
    if n==1:
        return 1
    else:
        return n*factorial(n-1)
```
<br><br>
Note how, if $n$ is not equal to 1, the function calls itself to calculate
the factorial of $n-1$.  This is recursion.  If we now say
`print(factorial(5))` the computer will correctly print the
answer 120.

* **a)** We encountered the Catalan numbers $C_n$ previously in Exercise 2.7
  on page 46.  With just a little rearrangement, the definition given there
  can be rewritten in the form
$$C_n = \left\lbrace\begin{array}{ll}
  \rule[-9pt]{0pt}{10pt}1 & \qquad\mbox{if $n=0$,} \\
  \dfrac{4n-2}{n+1}\,C_{n-1} & \qquad\mbox{if $n>0$.}
\end{array}\right.$$
Write a Python function, using recursion, that calculates $C_n$.  Use your
function to calculate and print $C_{100}$.



In [23]:
def C(n):
    if n==0:
        return 1
    elif n<0:
        return('error')
    else:
        return ((4*n-2)/(n+1)*C(n-1))

print("C(100)={0}".format(C(-3)))

C(100)=error


<br><br><br><br><br>

In [ ]:
#Part (a)
# Functions can be defined recursively!!!

def C(n):
    if n==0:
        return 1
    else:
        return ((4*n-2)/(n+1)*C(n-1))

print("C(100)=",C(100))

In [ ]:
#Part (a)
# Solution without recursive function
def C2(n):
    c = 1
    for i1 in range(1,n+1):
        c *= (4*i1-2)/(i1+1)
    return(c)
    
print("C2(100)=",C2(100))

<br><br><br><br>
* **b)** Euclid showed that the greatest common divisor $g(m,n)$ of two
  nonnegative integers $m$ and $n$ satisfies
$$g(m,n) = \biggl\lbrace\begin{array}{ll}
  m & \qquad\mbox{if $n=0$,} \\
  g(n,m\>\textrm{mod}\>n) & \qquad\mbox{if $n>0$.}
\end{array}$$
Write a Python function `g(m,n)` that employs recursion to calculate
the greatest common divisor of $m$ and $n$ using this formula.  Use your
function to calculate and print the greatest common divisor of 108 and 192.





<br><br><br><br><br>

In [ ]:
#Part (b)
# Recursive function solution

def gcd(m,n):
    if n==0:
        return m
    else:
        return gcd(n,m%n)

n,m = 108,192
print("The greatest common devisor of",n,"and",m,"is",gcd(108,192))

In [ ]:
#Part (b)
# Non-recursive solution

a,b = 108,192 
while b:
    # b is True as long as it is > 0
    print('a =',a,'\t b =',b,'\t a%b =',a%b)
    a,b = b, a%b
print('greatest common divisor =',a)

Comparing the calculation of the Catalan numbers in part (a) above with
that of Exercise 2.7, we see that it's possible to do the calculation two
ways, either directly or using recursion.  In most cases, if a quantity can
be calculated *without* recursion, then it will be faster to do so,
and we normally recommend taking this route if possible.  There are some
calculations, however, that are essentially impossible (or at least much
more difficult) without recursion.  We will see some examples later in this
book.

<br><br><br><br><br><br><br><br><br><br><br><br><br><br>

## Exercise:

Tetration may be thought of as the next operator after exponentiation: where $x\cdot n$ can be written as the sum $x+x+...+x$ with $n$ terms, and $x^{n}$ is the multiplication of $n$ factors: $x\cdot x\cdot x \cdot \cdots \cdot x$, the expression written $^{n}x$ is equal to the repeated exponentiation involving $n$ occurrences of $x$:

$$ \Large ^{n}x = x^{x^{x^{\cdots^{x}}}}$$

The exponentiation "tower" is evaluated from top to bottom.

Write a recursive function (a function that calls itself) to calculate $^{n}x$ and test it for small, positive real values of $x$ and non-negative integers $n$. (Tetration makes *very* large numbers!)

How many digits are there in $^{3}5$? In $^{5}2$?

<br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br>

### Pre-written solution

In [2]:
def tetration(n,base,top):
    """
    Sets top = base**top iteratively n times,
    then returns the final value of top.
    
    Thus tetration(n,x,x) computers x**x**... with n terms.
    """
    # If we've exhausted the tower, return the value    
    if n == 1:
        return top
    
    # Otherwise, compute the next exponential
    top = base ** top
    return tetration(n-1,base,top)
    

N = 5
x = 2
out = tetration(N,x,x)
print("n = {0}, x = {1} --> {2} characters".format(N,x,len(str(out))))
    
    
    
    
    

n = 5, x = 2 --> 19729 characters


<br><br><br><br><br><br><br><br><br><br><br><br>
## Programming Conventions

Your top priority when writing code is to get the code to work accurately. That said, some code will need to be used by you over the course of several years, or you may be writing code that someone else must be able to understand. There are several conventions that you should follow to keep your code easily readable by you and others:

* Be judicious in your use of comments.

* Use meaningful variable names.

* Use the right types of variables.

* Import functions at the top of your program.

* Give your constants names. Don't 'hardwire' numbers in your code.

* Employ user-defined functions, where appropriate.

* Print out partial results and updates throughout your program so that you and others know that long-running code isn't frozen or broken.

* Use a logical structure for your programs.

* Keep it simple! Don't make your programs unnecessarily complicated.

  + If you are writing code where speed is a high priority, sometimes you must value efficiency over readability. In all other cases, readability should be a top priority.

<br><br><br><br><br><br><br><br><br><br><br><br><br><br><br><br>
## Use pseudocode for planning code structure. This will be your 'narrative' on homework assignments.
* Video from Khan Academy: https://www.khanacademy.org/computing/computer-programming/programming/good-practices/pt/planning-with-pseudo-code
* Introduction: https://towardsdatascience.com/pseudocode-101-an-introduction-to-writing-good-pseudocode-1331cb855be7
* Examples: https://www.unf.edu/~broggio/cop2221/2221pseu.htm
* Reference: https://cs50.harvard.edu/ap/2020/assets/pdfs/pseudocode.pdf

